[merged] OstreeSePolicy: add ostree_sepolicy_get_csum()

jlebon opened the pull request:
This can be used as a fingerprint to determine whether two
OstreeSePolicy objects are equivalent.

Also add documentation for ostree_sepolicy_get_name().

cgwalters commented on src/libostree/libostree.sym:
```diff
@@ -321,4 +321,5 @@ global:
         ostree_sysroot_deployment_unlock;
         ostree_deployment_get_unlocked;
         ostree_deployment_unlocked_state_to_string;
+        ostree_sepolicy_get_csum;
```
This needs to be in the new 2016.5 section.

jlebon replied:
Hmm, did you forget to merge the bump commit into master?

cgwalters replied:
Weird...somehow, yes.  Anyways done now.

cgwalters commented on src/libostree/ostree-sepolicy.c:
```diff
@@ -257,6 +260,18 @@ initable_init (GInitable     *initable,
 
+      {
+        g_autofree guchar *csum_raw = NULL;
+        if (!ostree_checksum_file (etc_selinux_dir, OSTREE_OBJECT_TYPE_FILE,
```
Mmm...I would say use `ot_checksum_file` instead of this.  We don't actually care about uid/gid/perms or xattrs on the file I'd say...or more realistically the kernel and userspace selinux policy tools I'm almost sure don't.

As a bonus that function already returns an ASCII checksum.

jlebon replied:
Ahhh nice, will do.

cgwalters replied:
Also I'm a bit confused...we're checksumming `etc_selinux_dir`, the directory??

Also regarding policy and checksums, see http://marc.info/?l=selinux&m=145677940526857&w=2

jlebon replied:
Actually, `ot_checksum_file` only supports files, whereas `ostree_checksum_file` supports both.

jlebon replied:
Ahhh, I thought it was smart enough to descend in the directory and checksum the files. Hmm, OK, will do some more reading.

jlebon replied:
Right, it just did a checksum of the directory metadata, which is of course not what we want. Should have read more carefully.

I initially just thought checksumming the whole /etc/selinux dir was a safer & easier way to handle this, but it seems like that may be overkill and the only file that actually matters is the final compiled policy file, e.g. `policy.29`.

Should we just use `dirname(selinux_binary_policy_path())` and then checksum all the files under there (though there should only be one for the matching kernel it seems) ? Not sure yet if the path it returns respects `selinux_set_policy_root()` or if we'd have to build the full dirname ourselves.

cgwalters replied:
I think we need to enumerate files named `policy.[0-9]+` and pick the biggest.

cgwalters commented:
Regarding `selinux_binary_policy_path()`, indeed it should respect the root.  If you look at https://github.com/SELinuxProject/selinux/blob/master/libselinux/src/selinux_config.c#L222 it prefixes the internal paths with the root.

jlebon commented:
:arrow_up: Updated. It a bit more long-winded than expected :).

cgwalters commented on src/libostree/ostree-sepolicy.c:
```diff
@@ -155,6 +159,92 @@ ostree_sepolicy_class_init (OstreeSePolicyClass *klass)
+                                                         binfile_prefix);
+
+  g_autoptr(GFileEnumerator) direnum =
+    g_file_enumerate_children (bindir, OSTREE_GIO_FAST_QUERYINFO,
```
FWIW I've been trying to avoid using `GFile` for newer code but I know there's a lot of it left around here in the selinux bits.  I'm okay with this but we should keep in mind it should be ported later.

cgwalters commented on src/libostree/ostree-sepolicy.c:
```diff
@@ -155,6 +159,92 @@ ostree_sepolicy_class_init (OstreeSePolicyClass *klass)
+      GFileInfo *fi;
+      GFileType ftype;
+
+      if (!gs_file_enumerator_iterate (direnum, &fi, &child, cancellable, error))
```
This can actually just be `g_file_enumerator_iterate` now.

cgwalters commented on src/libostree/ostree-sepolicy.c:
```diff
@@ -155,6 +159,92 @@ ostree_sepolicy_class_init (OstreeSePolicyClass *klass)
+        {
+          g_autofree char *name = g_file_get_basename (child);
+
+          /* poor man's regex */
```
Well we do have `GRegex`...

jlebon replied:
Ahhh, will use.

cgwalters commented on src/libostree/ostree-sepolicy.c:
```diff
@@ -155,6 +159,92 @@ ostree_sepolicy_class_init (OstreeSePolicyClass *klass)
+
+          /* poor man's regex */
+          if (g_str_has_prefix (name, binfile_prefix) &&
+              fnmatch (binfile_no_fnmatch, name, FNM_NOESCAPE))
```
It confused me initially that `fnmatch` returns `0` if it matched.  Can we do `fnmatch (...) == FNM_NOMATCH` ?

jlebon replied:
Yes, it tricked as well. I'll just switch over to regex to make it easier.

cgwalters commented:
I'm okay with leaving the `GFile` stuff in (up to you) but it seems like using a regex would be better - we already have example usage of that in e.g. `_ostree_sysroot_parse_deploy_path_name()`

jlebon commented:
:arrow_up: Updated to use the glnx fd-based iterator.

cgwalters commented on src/libostree/ostree-sepolicy.c:
```diff
@@ -163,80 +163,85 @@ ostree_sepolicy_class_init (OstreeSePolicyClass *klass)
+  if (g_once_init_enter (&regex_initialized))
+    {
+      g_autofree char *sregex = g_strdup_printf ("^\\Q%s\\E\\.[0-9]+$",
+                                                 binfile_prefix);
```
This is a bit weird because we're basing a once-per-process value (the regex) on something that's determined dynamically.  Well actually it's always going to be `policy` right?  Eh.  Maybe this chunk would be better in `initable_init()` ?  Dunno.

It's okay as is too.

jlebon replied:
No I agree. I was coding everything else as if the prefix could be different from `policy`, so might as well go all the way here and make it per-instantiation. Since this is called from the initable_init, we could just leave it here and just not make it static, no? (Or maybe just use `g_regex_match_simple()` ?)

cgwalters replied:
Yeah, `g_regex_match_simple()` is totally fine here IMO.

cgwalters commented on src/libostree/ostree-sepolicy.c:
```diff
@@ -163,80 +163,85 @@ ostree_sepolicy_class_init (OstreeSePolicyClass *klass)
+  {
+    g_autofree char *path = glnx_fdrel_abspath (bindir_dfd, best_policy);
+    g_autoptr(GFile) file = g_file_new_for_path (path);
+    *out_csum = ot_checksum_file (file, G_CHECKSUM_SHA256, cancellable, error);
```
Here I'd say we should just change `ot_checksum_file` to take fd + path.  `git grep` says there are no other callers right now.  Instead of `g_file_read()`, we'd use `openat()` + `glnx_fd_readall_bytes()` ?

jlebon replied:
Agreed.

cgwalters replied:
Thanks; I know this patch has gone through a review wringer, but I believe it's worth it for future changes elsewhere so we can continue the fdrel transition.

cgwalters commented on src/libostree/ostree-sepolicy.c:
```diff
@@ -163,80 +163,85 @@ ostree_sepolicy_class_init (OstreeSePolicyClass *klass)
   g_autofree char *bindir_path = g_path_get_dirname (binary_policy_path);
-  g_autoptr(GFile) bindir = g_file_new_for_path (bindir_path);
-  g_autoptr(GFile) best_child = NULL;
+  g_autofree char *binfile_prefix = g_path_get_basename (binary_policy_path);
```
Also fwiw there's a non-malloc-ing variant of this in `glnx_basename()`.

jlebon commented:
:arrow_up: Rebased (though I didn't squash the fixups for easier review -- the first two commits are essentially the same). Changed to use `GRegex` and adapted `ot_checksum_file`.

cgwalters commented on src/libostree/ostree-sepolicy.c:
```diff
@@ -211,9 +197,18 @@ get_policy_checksum (char        **out_csum,
+           * compile-once GRegex... But picture how exciting it'd be if it *did*
+           * somehow change; there would be cheers & slow-mo high-fives at the
+           * sight of our code not breaking. Is that hope not worth a fraction
+           * of a millisecond? I believe it is... or maybe I'm just lazy. */
```
:laughing:

cgwalters commented:
@cgwalters-bot r+ 9f8b97f
